wrangler-core/src/main/java/io/cdap/wrangler/utils/OutputSchemaGenerator.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.wrangler.utils;
+
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.wrangler.api.Directive;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * This class can be used to generate the output schema after executing a set of directives. A list is maintained
+ * where each element is a map of fieldName --> value. Each element of the list corresponds to a directive in the list
+ * provided during initialization. Hence, each map represents the fields present across all output rows generated by a
+ * directive after execution.
+ */
+public class OutputSchemaGenerator {
```

Comment:
With the current design this class is very tightly coupled with `RecipePipelineExecutor` in fragile ways (e.g. using indexes to something passed in constructor is fragile.
I would move `generateDirectiveOutputSchema` back to  `RecipePipelineExecutor`. This way this class does not need `directives` and essentially becomes a fixed-size table of rows.

Or alternatively, make it store just a single directive with all the data and make `RecipePipelineExecutor` manage the list. It already has a list of `directives`, you can add another one or replace the original list with a new one

Reply:
Changed it to `DirectiveOutputSchemaGenerator` and handled the list separately